src/Compilers/CSharp/Portable/Symbols/PublicModel/EventSymbol.cs

```diff
@@ -62,6 +62,14 @@ ITypeSymbol IEventSymbol.Type
             }
         }
 
+        IFieldSymbol? IEventSymbol.BackingField
+        {
+            get
+            {
+                return _underlying.AssociatedField.GetPublicSymbol();
```

Comment:
>return _underlying.AssociatedField.GetPublicSymbol(); [](start = 16, length = 53)

It doesn't look like this implementation is going to work as advertised for events imported from metadata.

Reply:
I'm not opposed to making AssociatedField behave more uniformly when importing from metadata. On the other hand I think this is fine and we could change the advertising. There are two goals:

- Make it possible for analyzers to reason about object instance state, regardless of whether backing fields are included in GetMembers. Today, since event backing fields are not included in GetMembers, analyzers can't tell if an object holds state if it declares only events.
- Make it more convenient to go from a property to the matching backing field in GetMembers, if there is a backing field in GetMembers for the property.

I think both goals are met with the current implementation when considering both source properties and metadata properties.

Reply:
> On the other hand I think this is fine and we could change the advertising. 

Please provide clear specification for the new APIs and a set of targeted API tests that verify that they behave as specified.

---
In reply to: [532768881](https://github.com/dotnet/roslyn/pull/49659#discussion_r532768881) [](ancestors = 532768881)

Reply:
It seems like it could be less surprising and a matter of convenience for analyzer authors if Visual Studio heuristically paired backing fields in metadata back up if they fit past emit patterns. I'd be willing to put effort into this. Would that be a contribution you'd be willing to accept?

Reply:
> Make it possible for analyzers to reason about object instance state, regardless of whether backing fields are included in GetMembers.

In general, one cannot rely on presence of fields "to reason about object instance state" because, for example, private fields are not imported from metadata by default. 

---
In reply to: [532774220](https://github.com/dotnet/roslyn/pull/49659#discussion_r532774220) [](ancestors = 532774220,532768881)

Reply:
> In general, one cannot rely on presence of fields "to reason about object instance state" because, for example, private fields are not imported from metadata by default.

Thanks, that's a great point. That refocuses the whole first goal to be specifically about source properties.

> Please provide clear specification for the new APIs and a set of targeted API tests that verify that they behave as specified.

I will go ahead with this. What form should the specification be in, e.g. edited into the PR body at the top?

Reply:
> What form should the specification be in, e.g. edited into the PR body at the top?

This works for me. In particular, while I understand the reasoning behind adding these APIs, I'm somewhat concerned about the metadata vs source inconsistencies. I think we need a very clear understanding of what these APIs return, when they work, and when they do not.

Reply:
> That refocuses the whole first goal to be specifically about source properties.

I am not sure what exact conclusion you are making here. If one cannot get fields for imported symbols, why is it so important for source? Analyzers and other API consumers often deal with instances of types declared in metadata, therefore, we prefer to not expose APIs that do not provide consistent experience, especially if they are advertised as a way to detect something "very important". For example, we have IPropertySymbol.IsWithEvents property. It works consistently across source and metadata even though in order to calculate it for metadata we need to look at a private field, which is often not imported.

> What form should the specification be in, e.g. edited into the PR body at the top?

As the doc comment on the API(s) would be fine. Consumers of the API should be able to understand what to expect and should be able to relatively easily predict behavior of the API in different scenarios.

---
In reply to: [532787537](https://github.com/dotnet/roslyn/pull/49659#discussion_r532787537) [](ancestors = 532787537)

Reply:
I know it's important for source in a way that it's not important for metadata because I tried to write an analyzer to enforce that a class or struct declaration uses `.With` methods for each piece of state it contains.

> For example, we have IPropertySymbol.IsWithEvents property. It works consistently across source and metadata even though in order to calculate it for metadata we need to look at a private field, which is often not imported.

Good to know. I'd be interested in implementing a similar thing for IPropertySymbol.BackingField, even though it would behave like IsWithEvents in the presence of unimported metadata.

Reply:
> I'd be willing to put effort into this. Would that be a contribution you'd be willing to accept?

I think at this point it would be good to have complete proposal for the new APIs in the form of specification. Then we can evaluate it and make the decision.

---
In reply to: [532774783](https://github.com/dotnet/roslyn/pull/49659#discussion_r532774783) [](ancestors = 532774783)